src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -20,48 +22,41 @@ import scala.collection.mutable.ListBuffer
 import dotty.tools.dotc.core.Denotations.SingleDenotation
 import dotty.tools.dotc.core.SymDenotations.SymDenotation
 import dotty.tools.dotc.core.DenotTransformers.{SymTransformer, IdentityDenotTransformer, DenotTransformer}
+import Erasure.Boxing.adaptToType
 
-class LazyVals extends MiniPhaseTransform with SymTransformer {
+class LazyVals extends MiniPhaseTransform with IdentityDenotTransformer {
+  import LazyVals._
 
   import tpd._
 
-
-  def transformSym(d: SymDenotation)(implicit ctx: Context): SymDenotation = {
-    if (d is(Flags.Lazy, butNot = Flags.ModuleVal | Flags.Method)) {
-      // Method flag is set on lazy vals coming from Unpickler. They are already methods and shouldn't be transformed twice
-      d.copySymDenotation(
-        initFlags = d.flags | Flags.Method,
-        info = ExprType(d.info))
-    }
-    else d
-  }
-
   def transformer = new LazyVals
 
   val containerFlags = Flags.Synthetic | Flags.Mutable | Flags.Lazy
-  val initFlags = Flags.Synthetic | Flags.Method
+  val initFlags      = Flags.Synthetic | Flags.Method
+
+  val containerFlagsMask = Flags.Method | Flags.Lazy | Flags.Accessor | Flags.Module
 
   /** this map contains mutable state of transformation: OffsetDefs to be appended to companion object definitions,
     * and number of bits currently used */
   class OffsetInfo(var defs: List[Tree], var ord:Int)
-  val appendOffsetDefs = mutable.Map.empty[Name, OffsetInfo]
+  val appendOffsetDefs = mutable.Map.empty[Symbol, OffsetInfo]
 
   override def phaseName: String = "LazyVals"
 
-    /** List of names of phases that should have finished processing of tree
-      * before this phase starts processing same tree */
-    // override def ensureAfter: Set[String] = Set("mixin")
+  /** List of names of phases that should have finished processing of tree
+    * before this phase starts processing same tree */
+  override def runsAfter = Set(classOf[Mixin])
 
-    override def transformValDef(tree: ValDef)(implicit ctx: Context, info: TransformerInfo): Tree = {
-      if (!(tree.mods is Flags.Lazy) || (tree.mods is Flags.ModuleVal)) tree
+    override def transformDefDef(tree: DefDef)(implicit ctx: Context, info: TransformerInfo): Tree = {
+      if (!(tree.symbol is Flags.Lazy) || tree.symbol.owner.is(Flags.Trait)) tree
       else {
-        val isField = tree.symbol.owner.isClass
+          val isField = tree.symbol.owner.isClass
 
-        if (isField) {
-          if (tree.symbol.isVolatile) transformFieldValDefVolatile(tree)
-          else transformFieldValDefNonVolatile(tree)
-        }
-        else transformLocalValDef(tree)
+          if (isField) {
+            if (tree.symbol.isVolatile || tree.symbol.is(Flags.Module)) transformFieldValDefVolatile(tree)
```

Comment:
> Why should objects be treated as if they were always volatile?

That's the current semantics of objects. I'm fine to require the `@volatile` flag.

>We probably should make an exception for global objects 

What do you mean by global here? Static objects are not implemented here, they are handled by GenBCode itself.